### PR TITLE
WIP. Empty window titles cause errors. #183

### PR DESCRIPTION
**Pull request by BlitzKraft**
Added functions for testing empty titles and a config
window_title_empty.yaml.

### CODE REVIEW

**Comment by BlitzKraft** [comment]
I need some pointers on getting the window titles right. When the title is not set or empty, the title gets set to the shell (bash, in my case).

**Comment by tony** [comment]
can you rebase this against the latest codebase?

**Comment by BlitzKraft** [comment]
Will do.

**Comment by tony** [comment]
Sorry for the late response, I am having issues keeping up with the project. I'm seeking maintainers @ https://github.com/tony/tmuxp/issues/290. If you (or someone you know) may be interested in helping maintain libtmux or tmuxp, feel free to contact me by email or in the https://github.com/tony/tmuxp/issues/290 issue.

In the mean time, if you're still interested in this PR, can you rebase it via ``git pull --rebase origin master``?

**Comment by BlitzKraft** [comment]
I'll rebase and see how it goes. I am interested in maintaining this project.

**Comment by tony** [comment]
If you're not able to rebase immediately, that's also fine, but if you do, ping me here. GitHub has an issue notifying me when force pushes happen (even when I'm watching).

If you can pitch in time to help maintain, can you drop a message in #290? In addition to registering your interest, it would be encouraging  for more people to come forward, too. My hope is to always have at least one person on duty and a quick turnaround time on issues/PRs. I'm wondering if we need a couple people? Other than that, ATM I'm working on some issues to streamline contributions/issues/debugging.

**Comment by tony** [comment]
Also, I didn't mention this, but after a ``git pull --rebase origin master``, you will have to ``git push --force`` the branch. That's OK since it's a branch and a PR in this case.

**Comment by tony** [comment]
Can you give this a rebase?  @BlitzKraft

**Comment by tony** [comment]
@BlitzKraft Can you rebase?  (giving you a ping)

If I don't get a response, I may close the PR (I think?) - but you can still create a new PR if you / someone else regains interest in it

**Comment by BlitzKraft** [comment]
Got it, I will rebase.

**Comment by tony** [comment]
Nice!  That was quick. Welcome back @BlitzKraft

**Comment by BlitzKraft** [comment]
Seems I might have messed up. Not sure how to fix. What do you suggest?

**Comment by tony** [comment]
`git reflog` can help you

Since you probably have the old ref : https://stackoverflow.com/a/10099285/1396928

You'd `git reset --hard <ref>`

@BlitzKraft Would that help?

**Comment by BlitzKraft** [comment]
I think I got it. Did not reflog. I think this commit should be squashed. And I need to run the tests locally first. I keep forgetting to do that.

**Comment by tony** [comment]
@BlitzKraft Nice

Are the new gh action tests helpful?  They should detect if other functionality was affected / any tests added

**Comment by BlitzKraft** [comment]
I think they are helpful. I haven't used them very much. I need to get a hang of that.

**Comment by BlitzKraft** [comment]
`make test` encounters this: https://dpaste.org/zGVC 

Should I just ignore that?

**Comment by tony** [comment]
@BlitzKraft `poetry install -E "docs test coverage lint format"`

Does doing this and rerunning change anything?

`pip show pytest`

Are you using poetry for this? e.g. `poetry shell`, `poetry run py.test` ?

**Comment by tony** [comment]
(it's not a requirement but gives more piece of mind you have only the locked packages installed)

**Comment by BlitzKraft** [comment]
I was missing some dependencies. Installing them using poetry, like you mentioned, I was able to run all the tests. 

There is still one failing test. `test_window_title` - is there a separate test for empty titles?

File: tmuxp/workspacebuilder.py [hunk 1]
```diff
@@ -384,6 +384,9 @@ def pane_has_same_path(p):
         if all(pane_has_same_path(p) for p in w.panes):
             wconf['start_directory'] = w.panes[0].current_path
 
+        if w.name == '':
+            empty_window_title = True
```
Comment (tony) [inline tmuxp/workspacebuilder.py:388]:
```suggestion
        empty_window_title = w.name == ''
```

Assures the variable is always declared

Should fix https://github.com/tmux-python/tmuxp/pull/189/checks?check_run_id=991200510#step:15:372:

```
                wconf['panes'].append(pconf)
>               if empty_window_title:
E               UnboundLocalError: local variable 'empty_window_title' referenced before assignment

tmuxp/workspacebuilder.py:416: UnboundLocalError
```

File: tmuxp/config.py [hunk 2]
```diff
@@ -209,7 +209,10 @@ def expand(sconf, cwd=None, parent=None):
     if 'session_name' in sconf:
         sconf['session_name'] = expandshell(sconf['session_name'])
     if 'window_name' in sconf:
-        sconf['window_name'] = expandshell(sconf['window_name'])
+        if not (sconf['window_name'] is None):
```
Comment (tony) [inline tmuxp/config.py:212]:
```suggestion
        if sconf['window_name'] is not None:
```

Comment (BlitzKraft) [reply tmuxp/config.py:212]: This does not change the outcome of the test_window_title. As expected. The if statements - before and after the suggested changes - are equivalent. Were you expecting something different?

File: tmuxp/workspacebuilder.py (hunk 1 above)
Comment (BlitzKraft) [reply tmuxp/workspacebuilder.py:388]: That resolves the unbound variable. I am adding that change.

File: tmuxp/config.py (hunk 2 above)
Comment (BlitzKraft) [reply tmuxp/config.py:212]: Found something. `window_name` is being set to `tmux`. Hence the failing assertion. Now, need to dig where this is happening.

**Comment by tony** [comment]
@BlitzKraft Have you looked at the tests yet?

**Comment by tony** [comment]
@BlitzKraft 

If you want to continue with this contribution, please make a new pull request based off the newest master.

I am going to close this pull request for now as a significant amount of time has passed. 

Also I am at fault for letting the wait go too long in between.  Sorry about that.

## Window titles

This may behave differently depending on `tmux`'s config itself, via `automatic-rename` (e.g. `setw -g automatic-rename`)

A future PR would need to account for that

**Pull request closed without merging**